lib/elixir/src/elixir_lexical.erl

```diff
@@ -1,84 +1,118 @@
 %% Module responsible for tracking lexical information.
 -module(elixir_lexical).
--export([run/2, set_file/2, reset_file/1,
-  record_alias/4, record_alias/2,
-  record_import/4, record_import/5,
-  record_remote/3, record_struct/2,
-  format_error/1
-]).
+-export([run/2, with_file/3, trace/2, format_error/1]).
 -include("elixir.hrl").
 
 -define(tracker, 'Elixir.Kernel.LexicalTracker').
 
-run(File, Callback) ->
+run(#{tracers := Tracers} = E, Callback) ->
   case elixir_config:get(bootstrap) of
     false ->
       {ok, Pid} = ?tracker:start_link(),
-      try Callback(Pid) of
+
+      try Callback(E#{lexical_tracker := Pid, tracers := [?MODULE | Tracers]}) of
         Res ->
-          warn_unused_aliases(File, Pid),
-          warn_unused_imports(File, Pid),
+          warn_unused_aliases(Pid, E),
+          warn_unused_imports(Pid, E),
           Res
       after
         unlink(Pid),
         ?tracker:stop(Pid)
       end;
+
     true ->
-      Callback(nil)
+      Callback(E)
   end.
 
-%% RECORD
-
-record_alias(Module, Line, Warn, Ref) ->
-  if_tracker(Ref, fun(Pid) -> ?tracker:add_alias(Pid, Module, Line, Warn), ok end).
-
-record_import(Module, FAs, Line, Warn, Ref) ->
-  if_tracker(Ref, fun(Pid) -> ?tracker:add_import(Pid, Module, FAs, Line, Warn), ok end).
-
-record_alias(Module, Ref) ->
-  if_tracker(Ref, fun(Pid) -> ?tracker:alias_dispatch(Pid, Module), ok end).
-
-record_import(Module, Function, Arity, Ref) ->
-  if_tracker(Ref, fun(Pid) -> ?tracker:import_dispatch(Pid, Module, {Function, Arity}), ok end).
-
-record_remote(Module, EnvFunction, Ref) ->
-  if_tracker(Ref, fun(Pid) -> ?tracker:remote_dispatch(Pid, Module, mode(EnvFunction)), ok end).
+trace({import, Meta, Module, Opts}, #{lexical_tracker := Pid}) ->
+  {imported, Imported} = lists:keyfind(imported, 1, Meta),
+
+  Warn =
+    case lists:keyfind(warn, 1, Opts) of
+      {warn, false} -> false;
+      {warn, true} -> true;
+      false -> not lists:keymember(context, 1, Meta)
+    end,
+
+  Only =
+    case lists:keyfind(only, 1, Opts) of
+      {only, List} when is_list(List) -> List;
+      _ -> []
+    end,
+
+  ?tracker:add_import(Pid, Module, Only, ?line(Meta), Imported and Warn),
+  ok;
+trace({alias, Meta, _Old, New, Opts}, #{lexical_tracker := Pid}) ->
+  Warn =
```

Comment:
Maybe extract this to a `should_warn(Meta, Opts)`?

Reply:
Ah yes, good catch. I didn't notice it is reused because they were moved over from different places at different times!